File: server/routers/admin_auth.py
from fastapi import APIRouter, Cookie
from fastapi.responses import JSONResponse

from infra.log import logger
import services.state as state

router = APIRouter()


@router.get("/api/admin/auth/status")
async def admin_auth_status(admin_session: str | None = Cookie(default=None)):
    return {
        "need_setup": not await state.admin_auth.is_password_set(),
        "authenticated": await state.admin_auth.validate_session(admin_session),
    }


@router.post("/api/admin/auth/setup")
async def admin_auth_setup(body: dict):
    if await state.admin_auth.is_password_set():
        return JSONResponse({"error": "Password already set"}, status_code=400)
    password = body.get("password", "")
    if len(password) < 4:
        return JSONResponse({"error": "Password too short"}, status_code=400)
    await state.admin_auth.set_password(password)
    session = await state.admin_auth.create_session()
    logger.info("Admin password set up for the first time")
    resp = JSONResponse({"ok": True})
    resp.set_cookie(
        key="admin_session", value=session,
        httponly=True, path="/", samesite="lax", max_age=86400,
    )
    return resp


@router.post("/api/admin/auth/login")
async def admin_auth_login(body: dict):
    password = body.get("password", "")
    if not await state.admin_auth.verify_password(password):
        logger.warning("Admin login failed — wrong password")
        return JSONResponse({"error": "Wrong password"}, status_code=401)
    session = await state.admin_auth.create_session()
    logger.info("Admin logged in successfully")
    resp = JSONResponse({"ok": True})
    resp.set_cookie(
        key="admin_session", value=session,
        httponly=True, path="/", samesite="lax", max_age=86400,
    )
    return resp


@router.post("/api/admin/auth/logout")
async def admin_auth_logout(admin_session: str | None = Cookie(default=None)):
    await state.admin_auth.remove_session(admin_session)
    logger.info("Admin logged out")
    resp = JSONResponse({"ok": True})
    resp.delete_cookie(key="admin_session", path="/")
    return resp

File: backend/internal/router/admin_auth.go
package router

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/rs/zerolog/log"

	"astron-claw/backend/internal/middleware"
	"astron-claw/backend/internal/model"
)

func (app *App) adminAuthStatus(c *gin.Context) {
	ctx := c.Request.Context()
	adminSession, _ := c.Cookie("admin_session")

	needSetup := true
	isSet, err := app.AdminAuth.IsPasswordSet(ctx)
	if err == nil {
		needSetup = !isSet
	}

	authenticated := app.AdminAuth.ValidateSession(ctx, adminSession)

	c.JSON(200, gin.H{
		"code":          0,
		"need_setup":    needSetup,
		"authenticated": authenticated,
	})
}

func (app *App) adminAuthSetup(c *gin.Context) {
	ctx := c.Request.Context()

	isSet, err := app.AdminAuth.IsPasswordSet(ctx)
	if err != nil {
		log.Error().Err(err).Msg("Failed to check admin password status")
		middleware.MetricsErrorResponse(c, model.ErrChatInternalError)
		return
	}
	if isSet {
		middleware.MetricsErrorResponse(c, model.ErrAdminPasswordExists)
		return
	}

	var body struct {
		Password string `json:"password"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		middleware.MetricsErrorResponse(c, model.ErrChatInvalidReq)
		return
	}
	if len(body.Password) < 8 {
		middleware.MetricsErrorResponse(c, model.ErrAdminPasswordShort)
		return
	}

	if err := app.AdminAuth.SetPassword(ctx, body.Password); err != nil {
		log.Error().Err(err).Msg("Failed to set admin password")
		middleware.MetricsErrorResponse(c, model.ErrChatInternalError)
		return
	}

	session, err := app.AdminAuth.CreateSession(ctx)
	if err != nil {
		log.Error().Err(err).Msg("Failed to create admin session")
		middleware.MetricsErrorResponse(c, model.ErrChatInternalError)
		return
	}

	log.Info().Msg("Admin password set up for the first time")
	c.SetSameSite(http.SameSiteLaxMode)
	c.SetCookie("admin_session", session, 86400, "/", "", app.Config.Server.SecureCookie, true)
	c.JSON(200, gin.H{"code": 0})
}

func (app *App) adminAuthLogin(c *gin.Context) {
	ctx := c.Request.Context()

	var body struct {
		Password string `json:"password"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		middleware.MetricsErrorResponse(c, model.ErrChatInvalidReq)
		return
	}

	valid, err := app.AdminAuth.VerifyPassword(ctx, body.Password)
	if err != nil {
		log.Error().Err(err).Msg("Failed to verify admin password")
		middleware.MetricsErrorResponse(c, model.ErrChatInternalError)
		return
	}
	if !valid {
		log.Warn().Msg("Admin login failed — wrong password")
		middleware.MetricsErrorResponse(c, model.ErrAuthWrongPassword)
		return
	}

	session, err := app.AdminAuth.CreateSession(ctx)
	if err != nil {
		log.Error().Err(err).Msg("Failed to create admin session after login")
		middleware.MetricsErrorResponse(c, model.ErrChatInternalError)
		return
	}

	log.Info().Msg("Admin logged in successfully")
	c.SetSameSite(http.SameSiteLaxMode)
	c.SetCookie("admin_session", session, 86400, "/", "", app.Config.Server.SecureCookie, true)
	c.JSON(200, gin.H{"code": 0})
}

func (app *App) adminAuthLogout(c *gin.Context) {
	adminSession, _ := c.Cookie("admin_session")
	app.AdminAuth.RemoveSession(c.Request.Context(), adminSession)
	log.Info().Msg("Admin logged out")
	c.SetSameSite(http.SameSiteLaxMode)
	c.SetCookie("admin_session", "", -1, "/", "", app.Config.Server.SecureCookie, true)
	c.JSON(200, gin.H{"code": 0})
}
